src/node_locks.cc

```diff
@@ -0,0 +1,312 @@
+namespace worker {
+
+using v8::Array;
+using v8::Boolean;
+using v8::Context;
+using v8::Function;
+using v8::FunctionCallbackInfo;
+using v8::FunctionTemplate;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::Local;
+using v8::Null;
+using v8::Number;
+using v8::Object;
+using v8::Promise;
+using v8::String;
+using v8::TryCatch;
+using v8::Undefined;
+using v8::Value;
+
+LockManager LockManager::current_;
+
+void LockManager::SnapshotTask::AfterThreadPoolWork(int status) {
+  HandleScope scope(env_->isolate());
+  LockManager* manager = LockManager::GetCurrent();
+
+  Mutex::ScopedLock lock(manager->work_mutex_);
+
+  manager->Snapshot(env_, promise_.Get(env_->isolate()));
+}
+
+void LockRequest::AfterThreadPoolWork(int status) {
+  HandleScope scope(env_->isolate());
+  LockManager* manager = LockManager::GetCurrent();
+
+  Mutex::ScopedLock lock(manager->work_mutex_);
+
+  // If steal is true, then run these steps:
+  if (steal_) {
+    // For each lock of held:
+    for (auto const& lock : manager->held_locks_) {
+      // If lock's name is name, then run these steps:
+      if (lock->name() == name_) {
+        // Remove lock from held.
+        manager->held_locks_.erase(
+            std::remove(
+                manager->held_locks_.begin(), manager->held_locks_.end(), lock),
+            manager->held_locks_.end());
+
+        // Reject lock’s released promise with an "AbortError" DOMException.
+        Local<Function> domexception_ctor = env_->domexception_function();
+        CHECK(!domexception_ctor.IsEmpty());
+        Local<Value> argv[] = {
+          FIXED_ONE_BYTE_STRING(env_->isolate(), "Lock request could not be fulfilled"),
+          FIXED_ONE_BYTE_STRING(env_->isolate(), "AbortError"),
+        };
+        Local<Value> exception =
+            domexception_ctor
+                ->NewInstance(env_->context(), 2, argv)
+                .ToLocalChecked();
+        lock->released_promise()
+            ->Reject(env_->context(), exception)
+            .ToChecked();
+      }
+    }
+
+    // Prepend request in origin's lock request queue.
+    manager->pending_requests_.emplace_front(this);
+  } else {
+    // Otherwise, run these steps:
+    // If ifAvailable is true and request is not
+    // grantable, then run these steps:
+    if (if_available_ && !manager->IsGrantable(this)) {
+      // Let r be the result of invoking callback with null as the only
+      // argument. Note that r could be a regular completion, an abrupt
+      // completion, or an unresolved Promise.
+      Local<Promise::Resolver> promise = promise_.Get(env_->isolate());
+      TryCatch try_catch(env_->isolate());
+      Local<Value> r;
+      Local<Value> null = Null(env_->isolate());
+      if (callback_.Get(env_->isolate())
+              ->Call(env_->context(), Undefined(env_->isolate()), 1, &null)
+              .ToLocal(&r)) {
+        // Resolve promise with r and abort these steps.
+        promise->Resolve(env_->context(), r).ToChecked();
```

Comment:
@addaleax ... does this function need to include an `InternalCallbackScope`?

Reply:
I would say so, yes… That would also mean that this class should inherit from `AsyncWrap` as well (but that makes sense, since it does run JS code asynchronously), and ditto for the other place you just commented